### ISSUE TITLE
pipeline does not execute subsequent stacked commands after `ConnectionError`

### ISSUE THREAD

**Issue by leporid** [Grokzen/redis-py-cluster#176]
I ran into something similar to #168 after our Redis Cluster topology shifted and resulted in more network errors towards the clients, all in the context of pipelined commands.

In my case, I just decided to bump the redis-py-cluster version from 1.2.0 -> 1.3.3.  Now, there were no more exceptions in the client layer, but the application code saw problems with pipeline results.

Specifically, there were `None` values in the place of normal responses in the returned tuple from `pipeline.execute()`.  In my case, application code was expecting an iterable from a `zrange()` command.

I noticed that commands with `None` as a result are *not* considered in the re-attempt stack in `StrictRedisPipeline.send_cluster_commands()`.  The problem is that some commands may not have been run at all, and their `result` was never changed from `None`.  I believe I have addressed the problem with the following change:

    diff --git a/rediscluster/pipeline.py b/rediscluster/pipeline.py
    index 4fe4fb8..ea18d36 100644
    --- a/rediscluster/pipeline.py
    +++ b/rediscluster/pipeline.py
    @@ -199,7 +199,7 @@ class StrictClusterPipeline(StrictRedisCluster):
             # if we have more commands to attempt, we've run into problems.
             # collect all the commands we are allowed to retry.
             # (MOVED, ASK, or connection errors or timeout errors)
    -        attempt = sorted([c for c in attempt if isinstance(c.result, ERRORS_ALLOW_RETRY)], key=lambda x: x.position)
    +        attempt = sorted([c for c in attempt if isinstance(c.result, ERRORS_ALLOW_RETRY) or c.result is None], key=lambda x: x.position)
             if attempt and allow_redirections:
                 # RETRY MAGIC HAPPENS HERE!
                 # send these remaing comamnds one at a time using `execute_command`

I considered that there are cases in the [Redis protocol](https://redis.io/topics/protocol) where null responses are allowed for certain commands, and where `Connection.read_response()` will in fact return `NoneType`, but our application does not use timeout-able commands.

If this is a bug, and this fix makes sense, I would like to submit a very simple PR for it.  If there are more cases to consider (such as expected null responses) I'd be glad to dig in more.

Sincere Thanks!

**Comment by Grokzen** [comment]
Hi @leporid I would really like to fig into this one but i would like to get some more detailed info regarding steps to reproduce this bug. Do you have a script or some example chain of commands that is known to break? 

I think i understand the problem but unless i get some help to reproduce the problem i am a bit stuck right now.

**Comment by leporid** [comment]
Test case would be a pipeline where the first command reliably fails, then succeeds.  The subsequent commands must pertain to keys in the same slot as the first command.  Here is a test:

    from time import sleep
    from threading import Thread
    from rediscluster import StrictRedisCluster

    redis = StrictRedisCluster(startup_nodes=redis_nodes, socket_timeout=2)
    redis.delete('foo')

    pipeline = redis.pipeline()
    pipeline.blpop('foo', timeout=5)
    pipeline.set('foo', 'bar')
    pipeline.get('foo')

    def lpush():
        sleep(3)
        redis.lpush('foo', 'bar')

    thread = Thread(target=lpush)
    thread.start()

    result = pipeline.execute()
    print result

    thread.join()
    redis.delete('foo')

It will result in `[('foo', 'bar'), None, None]` with the given configuration, but if you change to `socket_timeout=10`, you get `('foo', 'bar'), True, 'bar']`.

[The change I made](https://github.com/leporid/redis-py-cluster/commit/f05c8413d4fa6bdb985d876f5c920e5d6bcf9c44) fixes this ensuring all commands are retried, but of course needs to be regression tested.  In practice, I have not had any problems.

Let me know if I need to go into more detail.

Cheers!

**Comment by myrfy001** [comment]
Hi, I am also meet this problem and I can provide another case for testing, here are the steps:
1) start up the redis cluster servers
2) use redis-py-cluster to connect to the servers
3) do something, make sure the client make a connection to the server
4) restart the redis cluster servers, by now, the connection is break, but the client seems doesn't known that happened~
5) user pipeline to execute some commands, like `smebers`
6) now, you are likely to have got `None` returned. you can buffer more commands if you don't get the `None` because I found that some buffered commands can be executed successfully but others not.

**Comment by Grokzen** [comment]
@72squared I would like your opinion on the matter and the fix that @leporid posted above here https://github.com/leporid/redis-py-cluster/commit/f05c8413d4fa6bdb985d876f5c920e5d6bcf9c44 I will try to dig into this and see if i can replicate the problem that @myrfy001 describes becuase that feels like a more reliable way to reproduce the problem if they are even related to each other.

**Comment by 72squared** [comment]
I thought the patch suggested by @myrfy001 looks okay at first glance, but after studying it, I think it might be problematic.

consider this: 
```
pipe = conn.pipeline()
pipe.get('foo') # where foo doesn't exist
pipe.execute()
```

It would return:

```
[None]
```

But with the patch suggested, it would have the side effect of always running that command twice. What we need to do is track down why some of the results are not being set with an error value in the result field. The NodeCommand class should be handling that but apparently not?

Here's another example that would have unintended side-effects:

```
lua_script = '''
redis.call('incr',  KEYS[1])
'''
pipe = c.pipeline()
pipe.eval(lua_script, 1, 'foo')
pipe.execute()
```
This would always return `[None]` but it would actually execute the incr call twice.

**Comment by Grokzen** [comment]
@72squared I would say that the threading case would be difficult to track down reliably because you know, threads... :) But if the error presents in ta similar way but through the way that @myrfy001 reported that it happens after the cluster servers restart it sounds more like it do not handle ASK/MOVED corectly more then anything else.

But at the same time it still feels like the client behaves properly in the case of a cluster restart. If the cluster restarts and all data i lost then it makes sense that it should return None for some commands because it would be the expected return value.

@myrfy001 Could you please explain some more what you mean by "buffered commands"?? I do not really get that point and how it relates or works in this case.

**Comment by leporid** [comment]
@72squared posed the question:
> The NodeCommand class should be handling that but apparently not?

To answer that question, I think there is an unintended early return in `NodeCommands` which exactly prevents the subsequent commands from being attributed with the first error.

[The exact modification to this early return is here](https://github.com/leporid/redis-py-cluster/commit/e2a33fac74e0d6e3313c755381b3474e5b6a302a), and it passes the first (thread-based) test I laid out above. You can probably disregard the previous patch I made, given it seems to have been a band-aid with the mentioned unwelcome side effects.

**Comment by myrfy001** [comment]
@Grokzen 

What I mean is put some command in the pipeline and the execute it.

```
            with rc.pipeline() as pipe:
                for key in ["a", "b", "c", "d"]:
                    pipe.smembers(key)
                cache_ret = pipe.execute()
```

now, if the key doesn't exist, retuen shoule be `[set([]), set([]), set([]), set([])]` ，but some times you will see something like this  `[set([]), None, set([]), set([])] `

**Comment by myrfy001** [comment]
By the way, I used the script tool provided redis' under `redis/utils/create-cluster`

I use the `./creata-cluster start` and `./creata-cluster create` to setup the environment.
Then, use `./creata-cluster stop` and `./creata-cluster start` to restart the cluster.
What i mean is I will stop all the servers at the same time and restart them again. I didn't test what will happen if I only stop a master node and restart it.

**Comment by myrfy001** [comment]
While debuging my project, another case happened and I can make sure it's because of ConnectionError.

If redis servers enables `timeout`, then the connection will be closed by server if it's not activate for some time. But the `redis-py-cluster` didn't expose this error to users and simply returns a None.

A new connection will be used for next request, so this will happen only once and makes it difficult to reproduce.

**Comment by ontholerian** [comment]
I think I found the issue. See line 443 in `rediscluster/pipeline.py`. I believe this `return` statement should be after the `for` loop (versus inside the loop), so that all commands get their results set to the exception value. 

In my test case, the first command of a two command pipeline was running into a `ConnectionError`, and I was able to verify that only the first command got the result set to the exception. What caused me confusion was that line 202 reattempted this command and so replaced that exception with the valid result, but it never reattempts the second command since it was left at `None`. HTH

**Comment by leporid** [comment]
I forked a while ago and [fixed exactly this](https://github.com/leporid/redis-py-cluster/commit/e2a33fac74e0d6e3313c755381b3474e5b6a302a).

**Comment by ontholerian** [comment]
Indeed, I didn't read the thread completely and missed that, apologies. Any reason why this isn't PR'd into here yet?

**Comment by leporid** [comment]
There is not.  Thank you for the suggestion.  #212

**Comment by hitigon** [comment]
I am encountering the same issue, ConnectioError on my side probably is caused by Redis Client Timeout as @myrfy001 pointed out. (2 seconds in my case, I am going to try different settings later). At first, I was testing with and without threads and multiprocessors, and turned out it's not related. The part of returning too early in read() as @ontholerian and @leporid mentioned, makes me confused as well, and I think @leporid 's fix should resolve the problem. Can we have the patch reviewed and merged? Thank you. @Grokzen

**Comment by Grokzen** [comment]
The fix in #212 has been merged into unstable branch. I am closing this one as fixed until it resurfaces again.
